discover/transport.go

```diff
@@ -0,0 +1,318 @@
+		if err != nil {
+			return nil, fmt.Errorf("Remote public key can't be "+
+				"parsed: %v", err)
+		}
+
+		txId, err := wire.NewShaHashFromStr(parts[4])
+		if err != nil {
+			return nil, fmt.Errorf("Can't decode 'txId' "+
+				"%v", err)
+		}
+
+		txPosition, err := strconv.Atoi(parts[6])
+		if err != nil {
+			return nil, fmt.Errorf("Can't decode 'index' "+
+				"%v", err)
+		}
+
+		return &ChannelAnnouncement{
+			FirstNodeID:  localPubKey,
+			SecondNodeID: remotePubKey,
+			TxId:         txId,
+			TxPosition:   txPosition,
+		}, nil
+
+	case "NODE":
+		if lenParts < 5 {
+			return nil, errors.New("Not enought arguments for " +
+				"'NODE' message")
+		}
+
+		hexPubKey, err := hex.DecodeString(parts[2])
+		if err != nil {
+			return nil, err
+		}
+
+		pubKey, err := btcec.ParsePubKey(hexPubKey, btcec.S256())
+		if err != nil {
+			return nil, fmt.Errorf("Public key can't be parsed:"+
+				" %v", err)
+		}
+
+		hostname := parts[3]
+		port, err := strconv.Atoi(parts[4])
+		if err != nil {
+			return nil, fmt.Errorf("Can't decode 'port': %v", err)
+		}
+
+		var alias string
+		if lenParts == 6 {
+			if len(parts[5]) > defaultMaxAliasLen {
+				return nil, errors.New("'alias' too long")
+			}
+			hexAlias, err := hex.DecodeString(parts[5])
+			if err != nil {
+				return nil, err
+			}
+			alias = string(hexAlias)
+		}
+
+		return &NodeAnnouncement{
+			NodeID: pubKey,
+			Ipv6:   hostname,
+			Port:   port,
+			Alias:  alias,
+		}, nil
+
+	}
+
+	return nil, errors.New("Unknown message format")
+}
+
+func (t *IRCTransport) Encode(a interface{}) (string, error) {
+	var parts []string
+
+	switch announce := a.(type) {
+	case *ChannelAnnouncement:
+		parts = make([]string, 10)
+		parts[1] = "CHAN"
+		parts[2] = hex.EncodeToString(announce.FirstNodeID.SerializeCompressed())
+		parts[3] = hex.EncodeToString(announce.SecondNodeID.SerializeCompressed())
+		parts[4] = announce.TxId.String()
+		parts[6] = strconv.Itoa(announce.TxPosition)
+
+	case *NodeAnnouncement:
+		parts = make([]string, 6)
+		parts[1] = "NODE"
+		parts[2] = hex.EncodeToString(t.pubKey.SerializeCompressed())
+		parts[3] = announce.Ipv6
+		parts[4] = strconv.Itoa(announce.Port)
+		parts[5] = hex.EncodeToString([]byte(announce.Alias))
+	default:
+		return "", errors.New("Unknown message format")
+	}
+
+	if rawSig, err := signMsg(t.privKey, parts); err != nil {
+		return "", err
+	} else {
+		parts[0] = hex.EncodeToString(rawSig)
+		return strings.Join(parts, defaultDataSeparator), nil
+	}
+}
+
+func (t *IRCTransport) Receive() chan interface{} {
+	return t.incoming
+}
+
+func (t *IRCTransport) Send(a interface{}) error {
+	if data, err := t.Encode(a); err != nil {
+		return err
+	} else {
+		t.conn.Privmsg(defaultLightningChannel, data)
+		return nil
+	}
+}
+
+func (t *IRCTransport) Wait() {
+	<-t.quit
+}
+
+func (t *IRCTransport) Joined() chan bool {
+	return t.joined
+}
+
+func (t *IRCTransport) onDisconnected(conn *client.Conn, line *client.Line) {
+	t.quit <- true
```

Comment:
Channel should be closed instead to be conformant with the rest of the codebase.